### ISSUE TITLE
How to evaluate a recurrent model using the EvalDLL

### ISSUE THREAD

**Issue by fwaris** [Microsoft/CNTK#1243]
I have a recurrent GRU model that apparently seems to train / test without errors using brainscript. I am trying to evaluate the model using the .net EvalDll but I can't seem to figure how to supply the input.

The model is for sequence classification (2-class).

This the model:
`

BrainScriptNetworkBuilder = [
	FDim=$FDim$
        LDim=$LDim$
        HDim = 20   		# hidden dimension

        t = DynamicAxis()
		features	= Input(FDim, dynamicAxis=t)
		labels  	= Input(LDim)

        R = BS.RNNs.RecurrentGRU(LDim, cellDim=HDim, features, inputDim=FDim).h
        O = BS.Sequences.Last(R)
        Y = CastAs(O,labels)

		#P    	= Softmax(zp)
		CE   	= CrossEntropyWithSoftmax(labels, Y)
		Err 	= ClassificationError(labels, Y)
		
		# Root Nodes
		featureNodes = (features)
		labelNodes = (labels)
		criterionNodes = (CE)
		evaluationNodes = (Err)
		outputNodes = (Y)
	]
`

The input is a) two dimension class label vector and b) 1 dimensional series of feature values under the 'swa' tag:

`

0 |swa 5 |l 1 0
0 |swa 5.0625
0 |swa 5.1875
0 |swa 5.1875
0 |swa 5.1875
0 |swa 5.25
0 |swa 5.375
0 |swa 5.5
0 |swa 5.6875
0 |swa 5.875
0 |swa 6.0625
0 |swa 6.25
0 |swa 6.5625
0 |swa 6.875
0 |swa 7.1875005
0 |swa 7.5
0 |swa 7.7499995
`
If I don't supply the labels in the input at evaluation time, cntk complains about label dimensions. If I supply the labels as 2 1-hot vectors, cntk complains that the number of labels do not match the number of the feature values in the input.

Please advise as to what is the correct way to evaluate a model such as this.

Thanks,

Faisal

**Comment by zhouwangzw** [comment]
Are you using IEvaluateModelManagedF()/IEvaluateModelManagedD() API? If yes, you can use GetNodeDimensions(NodeGroup.Input) to get all required inputs and their dimensions. You can find example [here](https://github.com/Microsoft/CNTK/blob/master/Examples/Evaluation/CSEvalClient/Program.cs).

**Comment by fwaris** [comment]
Zhou, thanks for the reply. Unfortunately the example does not show how to supply a **sequence** of 1D values as input. The new v2 API has Value.CreateSequence(...) but  I get a 'not implemented exception' when I call that.

**Comment by zhouwangzw** [comment]
@fwaris For sequence, you would better use EvalExtended API, IEvaluateModelExtended, which has better sequence support . The EvaluateExtendedNetworkSingleLayerNoInput() in [CSEvalClient](https://github.com/Microsoft/CNTK/blob/master/Examples/Evaluation/CSEvalClient/Program.cs) and [ManagedEvalTests ](https://github.com/Microsoft/CNTK/blob/master/Tests/UnitTests/ManagedEvalTests/EvalManagedTests.cs ) shows how to use the API. Regarding sequence, there is also an example in C++ [CppEvalExtendedClient] (https://github.com/Microsoft/CNTK/blob/master/Examples/Evaluation/CPPEvalExtendedClient/CPPEvalExtendedClient.cpp)  illustrating evaluation of the ATIS model using the Extended interface. Could you please have a try?

And the new V2 Eval C# API is still work-in-progress and some functions are not implemented yet. You can expect the V2 Eval C# API to be released very soon. 

Thanks.

**Comment by fwaris** [comment]
Thanks. I believe I get the gist. Here is a some sample code (F#) that shows how eval an RNN:
```fsharp
#r @"..\packages\Microsoft.Research.CNTK.CpuEval-mkl.2.0-beta7\lib\net45\x64\EvalWrapper.dll"
open Microsoft.MSR.CNTK.Extensibility.Managed
open System.IO
let modelPath = "model.dnn"

let toList (s:#seq<'a>) = new System.Collections.Generic.List<'a>(s)

let testEvalModel modelPath =
    let config = sprintf """ modelPath="%s" """ modelPath
    use mdl = new ModelEvaluationExtendedF()
    mdl.CreateNetwork(config)
    let inpScma = mdl.GetInputSchema()
    let outScma = mdl.GetOutputSchema()
    let outputVars = outScma |> Seq.map (fun v->v.Name) |> toList
    let inputBffr = inpScma.CreateBuffers<float32>(1,1) //two variable input: feature, label
    let outputBffr = outScma.CreateBuffers<float32>(1)
    mdl.StartForwardEvaluation(outputVars)
    let featureSeq = [for i in 1..50 -> float32 i] //test input sequence
    for feature in featureSeq do
        inputBffr.[0].Buffer.[0] <- feature //load input feature from sequence
        mdl.ForwardPass(inputBffr,outputBffr)
    outputBffr.[0].Buffer //return output buffer

testEvalModel modelPath
```
